core/pkg/artifacts/downloader.go

```diff
@@ -86,6 +88,42 @@ func (ad *ArtifactDownloader) getArtifactManifest(artifactID string) (manifest M
 	return manifest, nil
 }
 
+func (ad *ArtifactDownloader) getEntriesWithDownloadURLs(
+	artifactID string,
+	entriesToFetch []gql.ArtifactManifestEntryInput,
+	manifest Manifest,
+	manifestEntries map[string]ManifestEntry,
+) ([]ManifestEntry, error) {
+	var entries []ManifestEntry
+	response, err := gql.ArtifactFileURLsByManifestEntries(
+		ad.Ctx,
+		ad.GraphqlClient,
+		artifactID,
+		entriesToFetch,
+		manifest.StoragePolicyConfig.StorageLayout,
+		strconv.Itoa(int(manifest.Version)),
+	)
+	if err != nil {
+		return nil, err
+	}
+	for _, edge := range response.GetArtifact().GetFilesByManifestEntries().Edges {
+		filePath := edge.GetNode().Name
+		entry, err := manifest.GetManifestEntryFromArtifactFilePath(filePath)
+		if err != nil {
+			return nil, err
+		}
+		node := edge.GetNode()
+		if node == nil {
+			return nil, fmt.Errorf("error reading entry from fetched file urls")
+		}
+		entry.DownloadURL = &node.DirectUrl
+		entry.LocalPath = &filePath
+		entries = append(entries, entry)
+		delete(manifestEntries, filePath)
```

Comment:
if we send `x` entries to the query and the limit on the gorilla side is `y`, we will receive `min(x,y)` items. in order to make sure we don't delete entries that we actually aren't processing yet, we delete them from the manifest entries copy here.